File: Examples/GroupDocs.Viewer.Examples.CSharp/GroupDocs.Viewer.Examples/AzureInputDataHandler.cs
using GroupDocs.Viewer.Domain;
using GroupDocs.Viewer.Domain.Options;
using GroupDocs.Viewer.Handler.Input;
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.Auth;
using Microsoft.WindowsAzure.Storage.Blob;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace GroupDocs.Viewer.Examples.CSharp
{
    /// <summary>
    /// The implementation of GroupDocs.Viewer data handler for Azure Blob Storage.
    /// </summary>
    public class AzureInputDataHandler : IInputDataHandler
    {
        /// <summary>
        /// The blob delimiter.
        /// </summary>
        private const char Delimiter = '/';
        /// <summary>
        /// The endpoint template.
        /// </summary>
        private const string EndpointTemplate = "https://{account-name}.blob.core.windows.net/";
        /// <summary>
        /// The cloud blob container.
        /// </summary>
        private readonly CloudBlobContainer _container;
        /// <summary>
        /// Initializes a new instance of the <see cref="AzureInputDataHandler"/> class.
        /// </summary>
        /// <param name="accountName"></param>
        /// <param name="accountKey"></param>
        /// <param name="containerName"></param>
        public AzureInputDataHandler(string accountName, string accountKey, string containerName)
            : this(GetEndpoint(accountName), accountName, accountKey, containerName)
        {
        }
        /// <summary>
        /// Initializes a new instance of the <see cref="AzureInputDataHandler"/> class.
        /// </summary>
        /// <param name="endpoint">The endpoint e.g. https://youraccountname.blob.core.windows.net/ </param>
        /// <param name="accountName">The account name.</param>
        /// <param name="accountKey">The account key.</param>
        /// <param name="containerName">The container name.</param>
        public AzureInputDataHandler(Uri endpoint, string accountName, string accountKey, string containerName)
        {
            try
            {
                StorageCredentials storageCredentials = new StorageCredentials(accountName, accountKey);
                CloudStorageAccount cloudStorageAccount = new CloudStorageAccount(storageCredentials, endpoint, null, null, null);
                CloudBlobClient cloudBlobClient = cloudStorageAccount.CreateCloudBlobClient();
                TimeSpan? serverTimeout = cloudBlobClient.DefaultRequestOptions.ServerTimeout;
                cloudBlobClient.DefaultRequestOptions.ServerTimeout = TimeSpan.FromSeconds(10);
                _container = cloudBlobClient.GetContainerReference(containerName);
                _container.CreateIfNotExists();
                cloudBlobClient.DefaultRequestOptions.ServerTimeout = serverTimeout;
            }
            catch (StorageException e)
            {
                throw new System.Exception("Unable to recognize that Account Name/Account Key or container name is invalid.", e);
            }
            catch (FormatException e)
            {
                throw new System.Exception("Unable to recognize that Account Name/Account Key.", e);
            }
        }
        /// <summary>
        /// Gets the file description.
        /// </summary>
        /// <param name="guid">The unique identifier.</param>
        /// <returns>GroupDocs.Viewer.Domain.FileDescription.</returns>
        public FileDescription GetFileDescription(string guid)
        {
            try
            {
                string blobName = GetNormalizedBlobName(guid);
                CloudBlob blob = _container.GetBlobReference(blobName);
                blob.FetchAttributes();
                return new FileDescription(blobName, false)
                {
                    LastModificationDate = GetDateTimeOrEmptyDate(blob.Properties.LastModified),
                    Size = blob.Properties.Length
                };
            }
            catch (StorageException ex)
            {
                throw new System.Exception("Unabled to get file description.", ex);
            }
        }
        public Stream GetFile(string guid)
        {
            try
            {
                string blobName = GetNormalizedBlobName(guid);
                CloudBlob blob = _container.GetBlobReference(blobName);
                MemoryStream memoryStream = new MemoryStream();
                blob.DownloadToStream(memoryStream);
                memoryStream.Position = 0;
                return memoryStream;
            }
            catch (StorageException ex)
            {
                throw new System.Exception("Unabled to get file.", ex);
            }
        }
        /// <summary>
        /// Gets the last modification date.
        /// </summary>
        /// <param name="guid">The unique identifier.</param>
        /// <returns>System.DateTime.</returns>
        public DateTime GetLastModificationDate(string guid)
        {
            FileDescription fileDescription = GetFileDescription(guid);
            return fileDescription.LastModificationDate;
        }
        /// <summary>
        /// Gets files and folders for specified path.
        /// </summary>
        /// <param name="path">The path.</param>
        /// <returns>System.Collections.Generic.List&lt;GroupDocs.Viewer.Domain.FileDescription&gt;.</returns>
        public List<FileDescription> GetEntities(string path)
        {
            try
            {
                string normalizedPath = GetNormalizedBlobName(path);
                return GetFilesAndDirectories(normalizedPath);
            }
            catch (StorageException ex)
            {
                throw new System.Exception("Failed to load file tree.", ex);
            }
        }

        /// <summary>
        /// Gets the file tree.
        /// </summary>
        /// <param name="blobName">The blob name.</param>
        /// <returns>The file tree.</returns>
        private List<FileDescription> GetFilesAndDirectories(string blobName)
        {
            CloudBlobDirectory directory = _container.GetDirectoryReference(blobName);
            List<FileDescription> fileTree = new List<FileDescription>();
            foreach (IListBlobItem blob in directory.ListBlobs())
            {
                FileDescription fileDescription;
                CloudBlobDirectory blobDirectory = blob as CloudBlobDirectory;
                if (blobDirectory != null)
                {
                    fileDescription = new FileDescription(blobDirectory.Prefix, true);
                }
                else
                {
                    ICloudBlob blobFile = (ICloudBlob)blob;
                    fileDescription = new FileDescription(blobFile.Name, false)
                    {
                        Size = blobFile.Properties.Length,
                        LastModificationDate = GetDateTimeOrEmptyDate(blobFile.Properties.LastModified)
                    };
                }
                fileTree.Add(fileDescription);
            }
            return fileTree;
        }
        /// <summary>
        /// Gets normalized blob name, updates guid from dir\\file.ext to dir/file.ext
        /// </summary>
        /// <param name="guid">The unique identifier.</param>
        /// <returns>Normalized blob name.</returns>
        private string GetNormalizedBlobName(string guid)
        {
            return Regex.Replace(guid, @"\\+", Delimiter.ToString()).Trim(Delimiter);
        }

        /// <summary>
        /// Gets the endpoint e.g. https://youraccountname.blob.core.windows.net/
        /// </summary>
        /// <param name="accountName">The account name.</param>
        /// <returns>Endpoint Uri.</returns>
        private static Uri GetEndpoint(string accountName)
        {
            string endpoint = EndpointTemplate.Replace("{account-name}", accountName);
            return new Uri(endpoint);
        }
        /// <summary>
        /// Gets the file tree.
        /// </summary>
        /// <param name="blobName">The blob name.</param>
        /// <returns>The file tree.</returns>
        private List<FileDescription> GetFileTree(string blobName)
        {
            CloudBlobDirectory directory = _container.GetDirectoryReference(blobName);
            List<FileDescription> fileTree = new List<FileDescription>();
            foreach (IListBlobItem blob in directory.ListBlobs())
            {
                FileDescription fileDescription;
                CloudBlobDirectory blobDirectory = blob as CloudBlobDirectory;
                if (blobDirectory != null)
                {
                    fileDescription = new FileDescription(blobDirectory.Prefix, true);
                }
                else
                {
                    ICloudBlob blobFile = (ICloudBlob)blob;
                    fileDescription = new FileDescription(blobFile.Name, false)
                    {
                        Size = blobFile.Properties.Length,
                        LastModificationDate = GetDateTimeOrEmptyDate(blobFile.Properties.LastModified)
                    };
                }
                fileTree.Add(fileDescription);
            }
            return fileTree;
        }
        /// <summary>
        /// Gets date time or empty date.
        /// </summary>
        /// <param name="dateTimeOffset">The date time offset.</param>
        /// <returns>Date time or empty date.</returns>
        private DateTime GetDateTimeOrEmptyDate(DateTimeOffset? dateTimeOffset)
        {
            DateTime emptyDate = new DateTime(1, 1, 1);
            return dateTimeOffset.HasValue ? dateTimeOffset.Value.DateTime : emptyDate;
        }

        public void SaveDocument(CachedDocumentDescription description, Stream stream)
        {
            //TODO
        }
        /// <summary>
        /// Adds file to storage.
        /// </summary>
        /// <param name="guid">This is user defined key that identifies file in the storage.</param>
        /// <param name="content">Stream to save data to storage.</param>
        public void AddFile(string guid, Stream content)
        {
            try
            {
                string blobName = GetNormalizedBlobName(guid);
                ICloudBlob blob = _container.GetBlockBlobReference(blobName);
                blob.UploadFromStream(content);
            }
            catch (StorageException ex)
            {
                throw new System.Exception("Unabled to add file.", ex);
            }
        }

        public List<FileDescription> LoadFileTree(FileListOptions fileListOptions)
        {
            //TODO
            return new List<FileDescription>();
        }
        [Obsolete]
        public List<FileDescription> LoadFileTree(FileTreeOptions fileTreeOptions)
        {
            //TODO
            return new List<FileDescription>();
        }
    }
}

File: Examples/GroupDocs.Viewer.Examples.VisualBasic/AzureInputDataHandler.vb

Imports GroupDocs.Viewer.Domain
Imports GroupDocs.Viewer.Domain.Options
Imports GroupDocs.Viewer.Handler.Input
Imports Microsoft.WindowsAzure.Storage
Imports Microsoft.WindowsAzure.Storage.Auth
Imports Microsoft.WindowsAzure.Storage.Blob
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Text.RegularExpressions

Namespace GroupDocs.Viewer.Examples
    ''' <summary>
    ''' The implementation of GroupDocs.Viewer data handler for Azure Blob Storage.
    ''' </summary>
    Public Class AzureInputDataHandler
        Implements IInputDataHandler
        ''' <summary>
        ''' The blob delimiter.
        ''' </summary>
        Private Const Delimiter As Char = "/"c
        ''' <summary>
        ''' The endpoint template.
        ''' </summary>
        Private Const EndpointTemplate As String = "https://{account-name}.blob.core.windows.net/"
        ''' <summary>
        ''' The cloud blob container.
        ''' </summary>
        Private ReadOnly _container As CloudBlobContainer
        ''' <summary>
        ''' Initializes a new instance of the <see cref="AzureInputDataHandler"/> class.
        ''' </summary>
        ''' <param name="accountName"></param>
        ''' <param name="accountKey"></param>
        ''' <param name="containerName"></param>
        Public Sub New(accountName As String, accountKey As String, containerName As String)
            Me.New(GetEndpoint(accountName), accountName, accountKey, containerName)
        End Sub
        ''' <summary>
        ''' Initializes a new instance of the <see cref="AzureInputDataHandler"/> class.
        ''' </summary>
        ''' <param name="endpoint">The endpoint e.g. https://youraccountname.blob.core.windows.net/ </param>
        ''' <param name="accountName">The account name.</param>
        ''' <param name="accountKey">The account key.</param>
        ''' <param name="containerName">The container name.</param>
        Public Sub New(endpoint As Uri, accountName As String, accountKey As String, containerName As String)
            Try
                Dim storageCredentials As New StorageCredentials(accountName, accountKey)
                Dim cloudStorageAccount As New CloudStorageAccount(storageCredentials, endpoint, Nothing, Nothing, Nothing)
                Dim cloudBlobClient As CloudBlobClient = cloudStorageAccount.CreateCloudBlobClient()
                Dim serverTimeout As System.Nullable(Of TimeSpan) = cloudBlobClient.DefaultRequestOptions.ServerTimeout
                cloudBlobClient.DefaultRequestOptions.ServerTimeout = TimeSpan.FromSeconds(10)
                _container = cloudBlobClient.GetContainerReference(containerName)
                _container.CreateIfNotExists()
                cloudBlobClient.DefaultRequestOptions.ServerTimeout = serverTimeout
            Catch e As StorageException
                Throw New System.Exception("Unable to recognize that Account Name/Account Key or container name is invalid.", e)
            Catch e As FormatException
                Throw New System.Exception("Unable to recognize that Account Name/Account Key.", e)
            End Try
        End Sub
        
        ''' <summary>
        ''' Gets the file description.
        ''' </summary>
        ''' <param name="guid">The unique identifier.</param>
        ''' <returns>GroupDocs.Viewer.Domain.FileDescription.</returns>
        Public Function GetFileDescription(guid As String) As FileDescription Implements IInputDataHandler.GetFileDescription
            Try
                Dim blobName As String = GetNormalizedBlobName(guid)
                Dim blob As CloudBlob = _container.GetBlobReference(blobName)
                blob.FetchAttributes()
                Return New FileDescription(blobName, False) With { _
                    .LastModificationDate = GetDateTimeOrEmptyDate(blob.Properties.LastModified), _
                     .Size = blob.Properties.Length _
                }
            Catch ex As StorageException
                Throw New System.Exception("Unabled to get file description.", ex)
            End Try
        End Function
        Public Function GetFile(guid As String) As Stream Implements IInputDataHandler.GetFile
            Try
                Dim blobName As String = GetNormalizedBlobName(guid)
                Dim blob As CloudBlob = _container.GetBlobReference(blobName)
                Dim memoryStream As New MemoryStream()
                blob.DownloadToStream(memoryStream)
                memoryStream.Position = 0
                Return memoryStream
            Catch ex As StorageException
                Throw New System.Exception("Unabled to get file.", ex)
            End Try
        End Function
        ''' <summary>
        ''' Gets the last modification date.
        ''' </summary>
        ''' <param name="guid">The unique identifier.</param>
        ''' <returns>System.DateTime.</returns>
        Public Function GetLastModificationDate(guid As String) As DateTime Implements IInputDataHandler.GetLastModificationDate
            Dim fileDescription As FileDescription = GetFileDescription(guid)
            Return fileDescription.LastModificationDate
        End Function
       
        ''' <summary>
        ''' Gets the endpoint e.g. https://youraccountname.blob.core.windows.net/
        ''' </summary>
        ''' <param name="accountName">The account name.</param>
        ''' <returns>Endpoint Uri.</returns>
        Private Shared Function GetEndpoint(accountName As String) As Uri
            Dim endpoint As String = EndpointTemplate.Replace("{account-name}", accountName)
            Return New Uri(endpoint)
        End Function
        ''' <summary>
        ''' Gets the file tree.
        ''' </summary>
        ''' <param name="blobName">The blob name.</param>
        ''' <returns>The file tree.</returns>
        Private Function GetFileTree(blobName As String) As List(Of FileDescription)
            Dim directory As CloudBlobDirectory = _container.GetDirectoryReference(blobName)
            Dim fileTree As New List(Of FileDescription)()
            For Each blob As IListBlobItem In directory.ListBlobs()
                Dim fileDescription As FileDescription
                Dim blobDirectory As CloudBlobDirectory = TryCast(blob, CloudBlobDirectory)
                If blobDirectory IsNot Nothing Then
                    fileDescription = New FileDescription(blobDirectory.Prefix, True)
                Else
                    Dim blobFile As ICloudBlob = DirectCast(blob, ICloudBlob)
                    fileDescription = New FileDescription(blobFile.Name, False) With { _
                         .Size = blobFile.Properties.Length, _
                         .LastModificationDate = GetDateTimeOrEmptyDate(blobFile.Properties.LastModified) _
                    }
                End If
                fileTree.Add(fileDescription)
            Next
            Return fileTree
        End Function
        ''' <summary>
        ''' Gets normalized blob name, updates guid from dir\\file.ext to dir/file.ext
        ''' </summary>
        ''' <param name="guid">The unique identifier.</param>
        ''' <returns>Normalized blob name.</returns>
        Private Function GetNormalizedBlobName(guid As String) As String
            Return Regex.Replace(guid, "\\+", Delimiter.ToString()).Trim(Delimiter)
        End Function
        ''' <summary>
        ''' Gets date time or empty date.
        ''' </summary>
        ''' <param name="dateTimeOffset">The date time offset.</param>
        ''' <returns>Date time or empty date.</returns>
        Private Function GetDateTimeOrEmptyDate(dateTimeOffset As System.Nullable(Of DateTimeOffset)) As DateTime
            Dim emptyDate As New DateTime(1, 1, 1)
            Return If(dateTimeOffset.HasValue, dateTimeOffset.Value.DateTime, emptyDate)
        End Function
        Public Sub AddFile(guid As String, content As Stream) Implements IInputDataHandler.AddFile
            'TODO
        End Sub
        Public Function GetEntities(path As String) As List(Of FileDescription) Implements IInputDataHandler.GetEntities
            'TODO
            Return New List(Of FileDescription)()
        End Function
    End Class
End Namespace


